instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java

```diff
@@ -169,51 +185,200 @@ void captureMessage(LogRecordBuilder builder, AttributesBuilder attributes, Mess
 
     if (captureMapMessageAttributes) {
       // TODO (trask) this could be optimized in 2.9 and later by calling MapMessage.forEach()
-      mapMessage
-          .getData()
-          .forEach(
-              (key, value) -> {
-                if (value != null
-                    && (!checkSpecialMapMessageAttribute
-                        || !key.equals(SPECIAL_MAP_MESSAGE_ATTRIBUTE))) {
-                  attributes.put(getMapMessageAttributeKey(key), value.toString());
-                }
-              });
+      consumeAttributes(
+          mapMessage.getData()::forEach,
+          attributes,
+          checkSpecialMapMessageAttribute,
+          LogEventMapper::getMapMessageAttributeKey);
     }
   }
 
-  // visible for testing
-  void captureContextDataAttributes(AttributesBuilder attributes, T contextData) {
+  private static void consumeAttributes(
+      // Consumes an action on an entry, like map::forEach
+      Consumer<BiConsumer<String, Object>> entryActionConsumer,
+      ExtendedAttributesBuilder attributes,
+      boolean checkSpecialMapMessageAttribute,
+      BiFunction<String, ExtendedAttributeType, ExtendedAttributeKey<?>> keyProvider) {
+    entryActionConsumer.accept(
+        (key, value) -> {
+          if (value != null
+              && (!checkSpecialMapMessageAttribute || !key.equals(SPECIAL_MAP_MESSAGE_ATTRIBUTE))) {
+            consumeEntry(key, value, attributes, keyProvider);
+          }
+        });
+  }
+
+  @SuppressWarnings({"unchecked"})
+  private static void consumeEntry(
+      String key,
+      Object value,
+      ExtendedAttributesBuilder attributes,
+      BiFunction<String, ExtendedAttributeType, ExtendedAttributeKey<?>> keyProvider) {
+    if (value instanceof String) {
+      attributes.put(
+          (ExtendedAttributeKey<String>) keyProvider.apply(key, ExtendedAttributeType.STRING),
+          (String) value);
+    } else if (value instanceof Boolean) {
+      attributes.put(
+          (ExtendedAttributeKey<Boolean>) keyProvider.apply(key, ExtendedAttributeType.BOOLEAN),
+          (Boolean) value);
+    } else if ((value instanceof Long)) {
```

Comment:
elsewhere where we do something similar we also handle Byte and Short. You could use `value instanceof Byte
          || value instanceof Integer
          || value instanceof Long
          || value instanceof Short` and `((Number) value).longValue()` to reduce the number of `if` statements. The same strategy could also be applied to float and double.

Reply:
4ff3198ec34ef60a30ffe9baee83f8f865049293